xen/arch/arm/coproc/schedule.c

```diff
@@ -0,0 +1,317 @@
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return destroyed;
+}
+
+int vcoproc_scheduler_vcoproc_init(struct vcoproc_scheduler *sched,
+		struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	if ( !vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc) )
+		return -EINVAL;
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	vcoproc->sched_priv = VCOPROC_SCHED_OP(sched, alloc_vdata, vcoproc);
+	if ( !vcoproc->sched_priv )
+	{
+		printk("Failed to allocate scheduler-specific data for vcoproc \"%s\"\n",
+				dev_path(vcoproc->coproc->dev));
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return -ENOMEM;
+	}
+
+	vcoproc->state = VCOPROC_SLEEPING;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return 0;
+}
+
+int vcoproc_scheduler_vcoproc_destroy(struct vcoproc_scheduler *sched,
+		struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	if ( vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc) )
+		return 0;
+
+	vcoproc_sheduler_vcoproc_sleep(sched, vcoproc);
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state == VCOPROC_TERMINATING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return -EBUSY;
+	}
+
+	if ( vcoproc->sched_priv )
+	{
+		VCOPROC_SCHED_OP(sched, free_vdata, vcoproc->sched_priv);
+		vcoproc->sched_priv = NULL;
+	}
+	vcoproc->state = VCOPROC_UNKNOWN;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	return 0;
+}
+
+void vcoproc_sheduler_vcoproc_wake(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	/* TODO What to do if we came with state TERMINATING? */
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_SLEEPING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+
+	VCOPROC_SCHED_OP(sched, wake, vcoproc);
+	vcoproc->state = VCOPROC_WAITING;
+	spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+
+	vcoproc_schedule(sched);
+}
+
+void vcoproc_sheduler_vcoproc_sleep(struct vcoproc_scheduler *sched, struct vcoproc_instance *vcoproc)
+{
+	struct vcoproc_schedule_data *sched_data = sched->sched_priv;
+	unsigned long flags;
+	int reschedule = 0;
+
+	ASSERT(!vcoproc_scheduler_vcoproc_is_destroyed(sched, vcoproc));
+
+	spin_lock_irqsave(&sched_data->schedule_lock, flags);
+	if ( vcoproc->state != VCOPROC_WAITING && vcoproc->state != VCOPROC_RUNNING )
+	{
+		spin_unlock_irqrestore(&sched_data->schedule_lock, flags);
+		return;
+	}
+
+	VCOPROC_SCHED_OP(sched, sleep, vcoproc);
+	if ( vcoproc->state == VCOPROC_WAITING )
+		vcoproc->state = VCOPROC_SLEEPING;
+	else
+	{
+		vcoproc->state = VCOPROC_TERMINATING;
```

Comment:
VCOPROC_TERMINATING state means that currently running vcoproc (I mean vcoproc instance that currently runs on real coproc device) was asked to sleep. The running at the moment vcoproc might not go to sleep immediately as we want, so we can't set VCOPROC_SLEEPING state for it.